clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -0,0 +1,133 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+// Unit tests for cache replacement policies
+#include <iostream>
+#include <assert.h>
+#include "cache_replacement_policy_unit_test.h"
+#include "simulator/cache_lru.h"
+
+class cache_lru_test_t : public cache_lru_t {
+public:
+    void
+    initialize_cache(int associativity, int line_size, int total_size)
+    {
+        caching_device_stats_t *stats = new cache_stats_t(line_size, "", true);
+        if (!init(associativity, line_size, total_size, nullptr, stats, nullptr)) {
+            std::cerr << "LRU cache failed to initialize"
+                      << "\n";
+            exit(1);
+        }
+    }
+
+    void
+    unit_test_replace_which_way()
+    {
+        // Create and initialize a 4-way set associative cache with line size of 32 and
+        // total size of 256 bytes.
+        initialize_cache(4, 32, 256);
+
+        memref_t ref_1;
+        ref_1.data.type = TRACE_TYPE_READ;
+        ref_1.data.size = 1;
+        // Access the cache line in the following fashion. This sequence follows the
+        // sequence shown in https://github.com/DynamoRIO/dynamorio/issues/4881.
+        // Access the first row.
+        ref_1.data.addr = 0;
+        request(ref_1); // This accesses "a" in issue 4881.
+
+        ref_1.data.addr = 64;
+        request(ref_1); // This accesses "b" in issue 4881.
+
+        ref_1.data.addr = 128;
+        request(ref_1); // This accesses "c" in issue 4881.
+
+        ref_1.data.addr = 192;
+        request(ref_1); // This accesses "d" in issue 4881.
+
+        // After the above accesses, the counters for each way will be as follows:
+        //  way 0 ("a" in issue 4881): 3
+        //  way 1 ("b" in issue 4881): 2
+        //  way 2 ("c" in issue 4881): 1
+        //  way 3 ("d" in issue 4881): 0
+        // At this point way 0 ("a") has the highest counter so it should be replaced by
+        // the LRU policy.
+        assert(replace_which_way(0) == 0);
+
+        ref_1.data.addr = 0;
+        request(ref_1); // This replaces way 0 ("a").
+        // At this point way 0 has been replaced(accessed) and way 1 has the highest
+        // counter.
+        //  way 0 ("a" in issue 4881): 0
+        //  way 1 ("b" in issue 4881): 3
+        //  way 2 ("c" in issue 4881): 2
+        //  way 3 ("d" in issue 4881): 1
+        assert(replace_which_way(0) == 1);
+
+        ref_1.data.addr = 64;
+        request(ref_1); // This replaces way 1 ("b").
+        // At this point way 1 has been replaced(accessed) and way 2 has the highest
+        // counter.
+        //  way 0 ("a" in issue 4881): 1
+        //  way 1 ("b" in issue 4881): 0
+        //  way 2 ("c" in issue 4881): 3
+        //  way 3 ("d" in issue 4881): 2
+        assert(replace_which_way(0) == 2);
```

Comment:
suggest: Something more concise that makes it easy to add more test sequences?  What if a helper created a memref, called request, and checked that afterward the right way was considered the least recently used.  So it might look like this with lower-case showing which one is least recently used:

```C
access_and_check_lru(ADDRESS_A, 1); // A x X X
access_and_check_lru(ADDRESS_B, 2); // A B x X
access_and_check_lru(ADDRESS_C, 3); // A B C x
access_and_check_lru(ADDRESS_D, 0); // a B C D
access_and_check_lru(ADDRESS_A, 1); // A b C D
access_and_check_lru(ADDRESS_A, 1); // A b C D
access_and_check_lru(ADDRESS_A, 1); // A b C D
access_and_check_lru(ADDRESS_E, 2); // A E c D
```

Or if something else makes more sense to you.  I actually find these long comment sequences in the current code hard to scan and quickly see the state changes that matter and that something more concise is easier to read and step through mentally, and also seems to make it simpler to add a bunch more test sequences.

Reply:
Done. Added `access_and_check_lru` helper method.